micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsConnectionPoolListener.java

```diff
@@ -0,0 +1,114 @@
+/**
+ * Copyright 2019 Pivotal Software, Inc.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.core.instrument.binder.mongodb;
+
+import com.mongodb.MongoClient;
+import com.mongodb.connection.ServerId;
+import com.mongodb.event.*;
+import io.micrometer.core.instrument.Gauge;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.lang.NonNullApi;
+import io.micrometer.core.lang.NonNullFields;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * {@link ConnectionPoolListener} for collecting connection pool metrics from {@link MongoClient}.
+ *
+ * @author Christophe Bornet
+ */
+@NonNullApi
+@NonNullFields
+public class MongoMetricsConnectionPoolListener extends ConnectionPoolListenerAdapter {
+
+    private static final String METRIC_PREFIX = "mongodb.pool.";
+    private static final String POOLSIZE_METRIC_NAME = METRIC_PREFIX + "size";
+    private static final String CHECKEDOUT_METRIC_NAME = METRIC_PREFIX + "checkedout";
+    private static final String WAITQUEUESIZE_METRIC_NAME = METRIC_PREFIX + "waitqueuesize";
+
+    private final Map<ServerId, AtomicInteger> poolSize = new ConcurrentHashMap<>();
+    private final Map<ServerId, AtomicInteger> checkedOutCount = new ConcurrentHashMap<>();
+    private final Map<ServerId, AtomicInteger> waitQueueSize = new ConcurrentHashMap<>();
+
+    private final MeterRegistry registry;
+
+    public MongoMetricsConnectionPoolListener(MeterRegistry registry) {
+        this.registry = registry;
+    }
+
+    @Override
+    public void connectionPoolOpened(ConnectionPoolOpenedEvent event) {
+        registerGauge(event.getServerId(), POOLSIZE_METRIC_NAME, poolSize);
+        registerGauge(event.getServerId(), CHECKEDOUT_METRIC_NAME, checkedOutCount);
+        registerGauge(event.getServerId(), WAITQUEUESIZE_METRIC_NAME, waitQueueSize);
+    }
+
+    @Override
+    public void connectionPoolClosed(ConnectionPoolClosedEvent event) {
+        ServerId serverId = event.getServerId();
+        poolSize.remove(serverId);
```

Comment:
This looks like it will lead to a NPE next time the gauge is polled for a closed connection pool's gauges (line 107)
```java
p -> p.get(serverId).doubleValue()
```
Perhaps this should remove those gauges from the MeterRegistry instead.

Reply:
I'll check. I thought the Gauge would be automatically removed from the registry since the object it references doesn't exist anymore.

Reply:
@shakuzen there's indeed a NPE. I did a fix. Can you review ?